tests/OK/655b.lp

```diff
@@ -0,0 +1,22 @@
+symbol A:TYPE;
+symbol a:A;
+
+symbol +:A → A → A; notation + infix right 10;
+
+assert ⊢ a + + a a ≡ a + (a + a);
+assert ⊢ + a a + a ≡ (a + a) + a;
```

Comment:
I don't understand how it works, nor why we want such behavior.
In my mind, notation and underlying definition should be separated, this seems really confusing.

Reply:
I agree that it's confusing, probably we want to raise an error when two infix operators follow each other.

Reply:
In each of the example, it seems that one + is infix and the other is treated as a regular function.
I'd argue a symbol shouldn't be able to be treated as both infix and regular

Reply:
> ```
> assert ⊢ a + + a a ≡ a + (a + a);
> assert ⊢ + a a + a ≡ (a + a) + a;
> ```

As we can see, the behaviour is the following: when a binary operator is called at the beginning of an expression, or after another binary operator, it is not considered as a binary operator. It looks surprising and _ad-hoc_, but in fact in these positions, operators are called without left context, so it's not a bug. Whether it's a desirable feature is another question.

It may be more harmful than useful though.

Reply:
An application is a sequence of terms. When we have a sequence of terms, LP reorders its arguments according to the symbols and their priorities that are in these arguments, using the Pratt algorithm (see ref given by Gabriel).
@gabrielhdt should be able to explain us why we get this parsing, and whether it is possible to fail instead.

Reply:
It is indeed possible to fail. I'll update the parsing library.

Reply:
I created a separate issue https://github.com/Deducteam/lambdapi/issues/662 for this.